### PR TITLE
On Browser DOM scenarios use the SubtleCrypto API

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
See https://github.com/AaronRobinsonMSFT/SyncOverAsync for a design overview of the implementation here.

Added new `System.Security.Cryptography.Native.Browser` native library.

/cc @bartonjs @GrabYourPitchforks @lewing @jkoritzinsky @elinor-fung

### CODE REVIEW

**Comment by lewing** [comment]
For background:  This implements the platform hash functions on browser when both Subtle.crypto and SharedArrayBuffer are available.  This is currently true for Edge, Chrome and Firefox but untrue for Safari which doesn't currently support SharedArrayBuffer.

**Comment by kjpou1** [comment]
Just to open for discussion with the naming.  `crypto_worker.js` seems a little generic and may have a naming clash.  Something less generic and more obvious to identify the process would be `dotnet_crypto_worker.js`.

This is definitely not a showstopper.

**Comment by AaronRobinsonMSFT** [comment]
> Just to open for discussion with the naming. `crypto_worker.js` seems a little generic and may have a naming clash. Something less generic and more obvious to identify the process would be `dotnet_crypto_worker.js`.

That is a reasonable name. I will update to that.

**Comment by AaronRobinsonMSFT** [comment]
Failures in mono Windows x64 Debug - https://github.com/dotnet/runtime/issues/42677

**Comment by AaronRobinsonMSFT** [comment]
@lewing  @kg @vargaz @kjpou1 @bartonjs  The outstanding failure is unrelated to my changes - see previous comment. All the WASM legs have passed. This represents infrastructure for sync over async using a webworker and the SHA hashing support. Adding additional functionality should be straight forward now.

File: src/libraries/Native/AnyOS/System.Security.Cryptography.Native.Browser/pal_crypto_webworker.c [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+// Forward declarations
+extern int32_t dotnet_browser_sha_hash(
+    enum sha_hash ver,
+    uint8_t* input_buffer,
+    int32_t input_len,
+    uint8_t* output_buffer,
+    int32_t output_len);
+
+int32_t SystemCryptoNativeBrowser_SHAHash(
```
Comment (bartonjs) [inline src/libraries/Native/AnyOS/System.Security.Cryptography.Native.Browser/pal_crypto_webworker.c:15]: To avoid confusion of some digest algorithm being added that doesn't use the SHA name, I'd rename all of the "sha_hash" (various spellings thereof) to just "Digest", or "SimpleDigest" (since it's the simple one-shot, vs an accumulator/state-builder)

Comment (AaronRobinsonMSFT) [reply src/libraries/Native/AnyOS/System.Security.Cryptography.Native.Browser/pal_crypto_webworker.c:15]: I can do that. The reason I didn't is because SubtleCrypto only supports the SHA family - https://developer.mozilla.org/docs/Web/API/SubtleCrypto/digest. It would also require additional branching when dispatching in JavaScript based on the supplied enum.

Comment (AaronRobinsonMSFT) [reply src/libraries/Native/AnyOS/System.Security.Cryptography.Native.Browser/pal_crypto_webworker.c]: Done.

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs [hunk 2]
```diff
@@ -61,6 +61,9 @@ public static partial class PlatformDetection
         public static bool IsUsingLimitedCultures => !IsNotMobile;
         public static bool IsNotUsingLimitedCultures => IsNotMobile;
 
+        public static bool IsPlatformCryptoSupported => !IsBrowser || IsBrowserDomSupported;
+        public static bool IsNotPlatformCryptoSupported => !IsPlatformCryptoSupported;
```
Comment (bartonjs) [inline src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:65]: Is this ever true for real experiences, or just our current test harness/emulator?  (If it's true in real experiences, then having a test that shows we have some sort of sensible failure path seems virtuous)

**Review by bartonjs — APPROVED** [review]
The crypto parts LGTM.  The message pump makes conceptual sense to me, but I didn't give it a lot of scrutiny on the assumption that someone from runtime will.

**Comment by danmoseley** [comment]
Is there any security implications from using a web worker (passing unencrypted blob and key)? Presumably there's no trust boundary, eg in a web scenario each iframe has its own web worker, etc.

**Comment by AaronRobinsonMSFT** [comment]
>Is there any security implications from using a web worker (passing unencrypted blob and key)?

Yes there will be if a key is involved and will need to go through more review. This review is only for simple hashing so is okay, but if we start doing other things @bartonjs and company will need to scrutinize those uses.

Would you like some sort of comment?

**Comment by danmoseley** [comment]
Up to you. I was just curious.

**Comment by bartonjs** [comment]
Well, everything else in crypto uses keys... so it will come up on whatever the next thing is (assuming it's not asymmetric public-key-only).

In my mind the web worker didn't create a trust boundary, since the scope is (to my understanding) just the same browser tab.  If it has some other scope and is actually a trust boundary then we'll, of course, need to solve that problem.

The only thing I can think of doing is encrypt the channel, which we can't do, since we can't use a key without using the channel.  So I don't know it could work (even if something like JSON.stringify on an already-imported CryptoKey produces an identifier (vs the key) that can be used to reconnect on the other side, that feels like it has the same interception scope... but if it changes things like whether something got logged or temporarily written to disk then it would be a useful distinction).

**Comment by danmoseley** [comment]
Yes, it's almost surely fine, but someone expert in these things (@SteveSandersonMS ?) might know of some special precaution that we need to do.

File: src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/SHAHashProvider.Browser.cs [hunk 3]
```diff
@@ -62,965 +62,34 @@ public override int GetCurrentHash(Span<byte> destination)
...
-                int partInBase = ibStart;
-
-                /* Compute length of buffer */
-                bufferLen = (int)(_count & 0x3f);
-
-                /* Update number of bytes */
-                _count += partInLen;
-
-                fixed (uint* stateSHA256 = _stateSHA256)
+                fixed (byte* src = srcArray)
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/SHAHashProvider.Browser.cs:75]: Is the callee resilient against _src_ or _dest_ being nullptr? (This could occur if the source or destination are non-null 0-length buffers.)

**Review by GrabYourPitchforks — APPROVED** [review]
Overall LGTM. One question about the pinning logic. Somebody more familiar with the browser interop + JS layer will have to review that logic.

File: src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/SHAHashProvider.Browser.cs (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/SHAHashProvider.Browser.cs:75]: The nullptr `src` is tested. I will verify `dest`.

**Comment by GrabYourPitchforks** [comment]
@AaronRobinsonMSFT Just to double-check, `Module.HEAPU8` is accessible only from the DOM thread and not from any web worker thread, correct?

**Comment by AaronRobinsonMSFT** [comment]
>@AaronRobinsonMSFT Just to double-check, Module.HEAPU8 is accessible only from the DOM thread and not from any web worker thread, correct?

It is accessible from any WASM compiled component. However the memory it can access is limited to its own memory space _or_ what is shared through the [`SharedArrayBuffer`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer) API. If the WebWorker could access the DOM address space in question I could pass over the address and it would work. However since the WebWorker is isolated we need to use `SharedArrayBuffer`.

**Comment by GrabYourPitchforks** [comment]
@AaronRobinsonMSFT Thanks for the explanation! For context, my question wasn't really directed at your code. It was to help my understanding of what options might be available to us if we do need to throw key material over the wall to a web worker.

I think it's awesome that you got this infrastructure in place! I share your optimism that this will make future work more straightforward.

File: src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/SHAHashProvider.Browser.cs (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/SHAHashProvider.Browser.cs:75]: I've verified this is handled properly at the API layer.

**Comment by AaronRobinsonMSFT** [comment]
Failing run is because https://github.com/dotnet/runtime/issues/49213.

File: eng/SignCheckExclusionsFile.txt [hunk 4]
```diff
@@ -13,3 +13,4 @@
 *comhosttemplatecomhostdll.dll;;Template, DO-NOT-SIGN, https://github.com/dotnet/core-setup/pull/7549
 *staticapphosttemplateapphostexe.exe;;Template, DO-NOT-SIGN, https://github.com/dotnet/core-setup/pull/7549
 *dotnet.js;;Workaround, https://github.com/dotnet/core-eng/issues/9933
+*dotnet_crypto_worker.js;;Workaround, https://github.com/dotnet/core-eng/issues/9933
```
Comment (danmoseley) [inline eng/SignCheckExclusionsFile.txt:16]: @ViktorHofer just curious, what is the purpose of this file? From the linked issue, it seems that signing validation happens in an outer ring - in which case what is the purpose of these exclusions. Just curious

File: src/mono/wasm/runtime/library_channel.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+        constructor(msg_char_len) {
+            this.msg_char_len = msg_char_len;
+
+            const int_bytes = 4;
+            const comm_byte_len = int_bytes * (this.COMM_LAST_IDX + 1);
+            this.comm_buf = new SharedArrayBuffer(comm_byte_len);
+
+            // JavaScript character encoding is UTF-16.
+            const char_bytes = 2;
```
Comment (danmoseley) [inline src/mono/wasm/runtime/library_channel.js:28]: No idea if it is relevant, but I wonder whether we have tests that hash characters out of the BMP, ie., that decode to more than 2 bytes ?

Comment (AaronRobinsonMSFT) [reply src/mono/wasm/runtime/library_channel.js:28]: > No idea if it is relevant, but I wonder whether we have tests that hash characters out of the BMP, ie., that decode to more than 2 bytes ?

@danmoseley It shouldn't matter here as this is merely passing around a JSONified object and if it contains strings those will be encoded as UTF-16 which supports the 17 planes - it may use 2 16-bit characters instead of 1.

The second part, regarding hashing of characters outside of BMP, should exist somewhere, although I am unclear if it matters much here or if it occurs in any of the crypto hash testing at all.

Comment (danmoseley) [reply src/mono/wasm/runtime/library_channel.js:28]: Got it - fine.

File: eng/SignCheckExclusionsFile.txt (hunk 4 above)
Comment (ViktorHofer) [reply eng/SignCheckExclusionsFile.txt:16]: That file predates the outer ring signing validation. @mmitche do you know if the Validate-DotNet outer ring validation honors the SignCheckExclusionFile in our repo?

Comment (ViktorHofer) [reply eng/SignCheckExclusionsFile.txt:16]: Also just saw that the outer ring signing validation is failing and opened https://github.com/dotnet/core-eng/issues/12601 to better understand who tracks that pipeline.

**Comment by SteveSandersonMS** [comment]
This makes sense and looks good to me! Two main questions:

### Locking the main thread

At first I was a bit nervous about the spin-lock blocking the main thread, since it's something web devs are generally told never to do, but in this case I understand we can trust that the worker will complete its task very quickly and almost deterministically.

What's the longest you think we might be spin-locking the main thread for? If the user is trying to hash a few megabytes, is that still likely to complete in < 100ms, even on a 5-year-old phone? If we risk getting into the multi-seconds territory then browsers will start showing the "Page is not responding" prompt or will forcibly reload the page on some mobile OS cases.

### Browser support

I understand that Safari doesn't enable `SharedArrayBuffer` by default yet, and we have no way to know when or if they will do. Perhaps I should be able to tell this from the code, but I didn't spot an obvious answer: is there anything in the design here that forces developers to explicitly opt into using a feature that doesn't work on some browsers that we explicitly support?

That is, we include Safari in [our list of supported browsers](https://docs.microsoft.com/en-us/aspnet/core/blazor/supported-platforms?view=aspnetcore-5.0), so the normal promise we're making is that all the APIs we offer will work (and produce the same results) on all those supported browsers. Developers might only test locally on Edge/Chrome, and will trust us that it's also going to run on Safari on macOS or iOS. Is there a risk we're setting them up for a bad surprise in production? I think these APIs would be the first examples of that.

We might want to do something to force developers to opt into this feature, so they have to realise they are reducing their browser support surface. I'm not sure how we'd achieve that technically without doing so with TFMs. Maybe it would be enough to do something with the Platform Compatibility Analyzer (even though it wouldn't be able to spot usages in NuGet packages you're consuming). cc @danroth27 for any thoughts on developer experience and how we communicate the browser support limitation.

**Comment by AaronRobinsonMSFT** [comment]
>If we risk getting into the multi-seconds territory then browsers will start showing the "Page is not responding" prompt or will forcibly reload the page on some mobile OS cases.

@SteveSandersonMS Great question. I'm not sure the best way to mitigate that concern. We can have a "timeout" but then how does one say - it is okay to run longer? We have no "retry" semantic here. Perhaps we could have a callback in a new API?

> Perhaps I should be able to tell this from the code, but I didn't spot an obvious answer: is there anything in the design here that forces developers to explicitly opt into using a feature that doesn't work on some browsers that we explicitly support?

Not at present. That is something we need to validate about how much support we need for non-`SharedArrayBuffer` enabled systems. @lewing is following up on requirements here.

**Comment by danroth27** [comment]
>> Perhaps I should be able to tell this from the code, but I didn't spot an obvious answer: is there anything in the design here that forces developers to explicitly opt into using a feature that doesn't work on some browsers that we explicitly support?
>
> Not at present. That is something we need to validate about how much support we need for non-SharedArrayBuffer enabled systems. @lewing is following up on requirements here.

This sounds like it might be a blocking issue then. What would the user experience be on Safari if you made this change? PlatformNotSupportedException?

**Comment by AaronRobinsonMSFT** [comment]
>What would the user experience be on Safari if you made this change? PlatformNotSupportedException?

Yes. https://github.com/dotnet/runtime/blob/c749ab5692cdddfc8e4bebc588dfacc94e26b0fb/src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/SHAHashProvider.Browser.cs#L78-L82

**Comment by SteveSandersonMS** [comment]
>> If we risk getting into the multi-seconds territory then browsers will start showing the "Page is not responding" prompt or will forcibly reload the page on some mobile OS cases.
>
> @SteveSandersonMS Great question. I'm not sure the best way to mitigate that concern. We can have a "timeout" but then how does one say - it is okay to run longer? We have no "retry" semantic here. Perhaps we could have a callback in a new API?

This is definitely tricky and gets to the core of the challenge of doing sync in an environment that's hostile to sync. Maybe you have a solution in mind, but I don't see how a callback would change the picture. Either we yield the main execution thread (in which case it's not a single sync call any more) or we don't (in which case we run the risk of killing the app). The options I can anticipate are:

 1. We do nothing, and let the main thread stay blocked as long as it takes. The drawback is that, if it's too long, then Chrome will show the "Web page is not responsive - kill it?" prompt, and iOS Safari will go ahead and kill it without even asking. This is pretty dangerous if the semantics of the call are meant to be that it's allowed to take a full second or more for reasonable data sizes.
 2. Or, we impose a short timeout of our own (say, 100ms) and throw a timeout exception if the processing takes longer. This is safer, in that the developer has a way to catch the error and the app won't be killed, but it remains complicated to use the API if you don't know the data sizes or speed of the client device up front.
 3. Or, we could simulate blocking via re-entrancy (same as the "asyncify" tool you linked in the design doc). From the .NET side, it could be made to look indistinguishable from actual blocking, in that the call stack could be recovered exactly on restart. The one aspect of the semantics that couldn't be made perfect is synchronous JS-to-.NET calls, which would fail to be synchronous from the JS perspective if they involved calling one of these APIs. Maybe we could argue that's close enough.

Is the callback option something that asks the developer every 100ms whether they want to continue waiting? That may help a bit but if the developer just returns `true` all the time, the app is still at risk of being killed after a few seconds.

**Comment by AaronRobinsonMSFT** [comment]
>Is the callback option something that asks the developer every 100ms whether they want to continue waiting? That may help a bit but if the developer just returns true all the time, the app is still at risk of being killed after a few seconds.

@SteveSandersonMS Yep. I wasn't suggesting returning for all the reasons mentioned. The callback would be called by the main thread after so much time and allow the app doing the hash to make a decision and provide UX to say "The payload was too big". It doesn't help the end user in any way though because as mentioned it requires the app developer to provide a reasonable experience.

Ideally (3) is the solution, https://github.com/dotnet/runtime/issues/48713, but if the callback is too much then (2) is something we could deliver. That does however force us into policy decisions I'm not convinced we can solve any better than the app developer.

**Comment by kg** [comment]
> What's the longest you think we might be spin-locking the main thread for? If the user is trying to hash a few megabytes, is that still likely to complete in < 100ms, even on a 5-year-old phone? If we risk getting into the multi-seconds territory then browsers will start showing the "Page is not responding" prompt or will forcibly reload the page on some mobile OS cases.

While crypto is pretty fast (and even hardware accelerated on most modern CPUs) we're dealing with web workers here so you should assume that it could stall for multiple seconds. A few things to consider:

* Browsers may set their web worker threads to low priority to stop websites from degrading the performance of the system as a whole
* AFAIK, the spec allows browsers to set a limit on the number of threads used for web workers, so we could end up sharing a worker thread with other webpages
* The host machine may not have enough cores to handle our workload (browser thread + crypto thread) alongside the other things competing for their machine (8 chrome processes for gmail and twitter and youtube tabs, 500 electron processes)
* The spec allows browsers to kill a worker at any time, which could halt the crypto operation (and if we're spinning waiting for a response message, we'll hang forever). We could probably fix this by probing the status of the Worker object though, I think.
* I vaguely recall the spec at one point saying that Workers don't even ever have to start (i.e. the browser can decide There Are Too Many Workers and just not start yours until there's "room") but I can't find the language specifying this in the w3 spec right now. We should make sure that we don't hang the main thread in that scenario either

A 'should I keep running' callback provided by the user sounds like a good solution for some of these problems.

**Comment by SteveSandersonMS** [comment]
Thanks for all the additional context - that’s valuable. 

> A 'should I keep running' callback provided by the user sounds like a good solution for some of these problems.

To clarify whether it would be a good solution, what logic would we recommend people put into such a callback? Do we recommend a specific timeout? Is there a scenario where it would make sense for them to have more complex rules, and if so, what examples?

**Comment by kg** [comment]
> Thanks for all the additional context - that’s valuable.
> 
> > A 'should I keep running' callback provided by the user sounds like a good solution for some of these problems.
> 
> To make that more concrete, what logic would we recommend people put into such a callback? Do we recommend a specific timeout? Is there a scenario where it would make sense for them to have more complex rules, and if so, what examples?

Simple is probably best, but (for example) the app developer knows whether they're building a batch data processing tool ('create archive', 'export document', etc) vs a user-facing app that should be responsive and low-latency. In the latter case they might want to abort on the very first call with a "try again" message or something, while in the former they probably want to keep going

It's too bad we don't have any way to provide them with a progress value so they can tell whether the operation is stalled.

**Comment by AaronRobinsonMSFT** [comment]
This callback feature is going to require an API proposal and review.

**Comment by SteveSandersonMS** [comment]
I know this is a total pain, but it would be great to have a sense of what sorts of data sizes can realistically be processed on low-end hardware via these APIs before the operation would have to be aborted (or the app typically gets killed). This would help to clarify whether the timeout concerns are just a theoretical edge case we can ignore, or whether this approach would really solve customer problems or create new ones.

When we decided to implement these APIs for WebAssembly, was this based on any particular customer requests, and is there any info about what the scenarios were (and hence how much data they are likely to be hashing/encrypting/decrypting)?

Another possible approach we could consider (which would also avoid the Safari problem) is supporting a mode where *all* the .NET code runs in a worker thread - [something we considered very early on](https://github.com/dotnet/aspnetcore/issues/5475) but didn't pursue due to lack of a reason. This would complicate numerous things for Blazor, but for customers who are willing to give up on unmarshalled interop, would mean we could freely block the .NET execution any time we want. Of course, the asyncify solution would be even more flexible.

To be clear, I really don't have any agenda to derail your solution here. If it works with the realistic customer scenarios we have in mind and are OK with the Safari limitation, then it's much easier than changing how the whole runtime is hosted!

**Comment by danmoseley** [comment]
> Or, we could simulate blocking via re-entrancy (same as the "asyncify" tool you linked in the design doc).

Just for my education, how would this be done? It sounds like "pumping messages" in Windows terms. Perhaps I'm forgetting how one does this in Javascript.

**Comment by SteveSandersonMS** [comment]
> Just for my education, how would this be done? It sounds like "pumping messages" in Windows terms. Perhaps I'm forgetting how one does this in Javascript.

Yes, more or less. The .NET interpreter could return from its invocation before finishing the code it's interpreting through, keeping track of all the state it needs to resume from that instruction on a subsequent call in from JS. As far as the .NET code is concerned, the universe was frozen for a while (like being blocked) but no other difference in behavior is observable. We actually did this on the original DotNetAnywhere runtime (before the migration to Mono WebAssembly) in order to support some sync calls.

However JavaScript code *would* be able to observe this happening. Synchronous (blocking) calls from JS to .NET would cease to be blocking if any part of the callee's operations involved one of these universe-suspending phases. We would need some way to make this an error, i.e., when calling from JS to .NET we'd have to know whether the caller is expecting to get back a promise or not, and if they are not that means they want the call to be synchronous, and therefore any suspension becomes an exception. If the caller is getting back a promise, then we could so some stuff in JS to make the universe-suspended-phase part so the JS promise so the caller can't tell the difference.

In summary, it's complicated and still has edge cases. It *might* be the best available solution even so.

**Comment by AaronRobinsonMSFT** [comment]
> In summary, it's complicated and still has edge cases. It might be the best available solution even so.

Agree. It could also be done in vNext if the current solution is an acceptable stop-gap for the current level of support.

Also I would very much recommend going down the path of adding support for Asyncify, https://github.com/dotnet/runtime/issues/48713, rather than doing some bespoke for Blazor.

**Comment by bartonjs** [comment]
> we'd have to know whether the caller is expecting to get back a promise or not, and if they are not that means they want the call to be synchronous, and therefore any suspension becomes an exception

I don't know how the JS -> .NET calls work, but would it make sense to just make them all be promises in that case?  Sure, it's how we got into this problem in the first place, but it's sort of a "when in Rome..."

**Comment by SteveSandersonMS** [comment]
We already support both synchronous and asynchronous interop APIs. Both are valuable for different use cases.

**Comment by marek-safar** [comment]
@SteveSandersonMS we'll also support "proper" async crypto in the form of new APIs. We discussed that with Jeremy to look into if it's worth having that as a generic API or whether we will add it as browser-specific. This has lower priority because we expect the usage to be lower than existing crypto APIs but it'd be the right options for those targetting the primary web.

/cc @lewing

**Comment by lewing** [comment]
@AaronRobinsonMSFT I'm happy to resolve the conflict or something like should fine for now.

```xml
     <_DotnetJSSrcWorkerFile Include="$(_WasmRuntimePackSrcDir)pal_crypto_webworker.js" />
     <_DotnetJSSrcPreFile Include="$(_WasmRuntimePackSrcDir)library_channel.js" />

     <_DotnetJSSrcFile Include="$(_WasmRuntimePackSrcDir)\*.js" />
     <_DotnetJSSrcFile Remove="@(_DotnetJSSrcWorkerFile)" />
     <_DotnetJSSrcFile Remove="@(_DotnetJSSrcPreFile)" />
```

**Comment by SteveSandersonMS** [comment]
@lewing I see you're doing some work to make this PR build pass, but have we reached a conclusion on whether it should be merged?

The last I understood (as per [this comment](https://github.com/dotnet/runtime/pull/49511#issuecomment-811294074)) was that we saw it would be necessary to do some more research to determine the realistic scenarios and find out whether the spin-blocking was actually viable for those scenarios, or whether it would just result in people's apps being terminated by the browser.

**Comment by lewing** [comment]
> @lewing I see you're doing some work to make this PR build pass, but have we reached a conclusion on whether it should be merged?
> 
> The last I understood (as per [this comment](https://github.com/dotnet/runtime/pull/49511#issuecomment-811294074)) was that we saw it would be necessary to do some more research to determine the realistic scenarios and find out whether the spin-blocking was actually viable for those scenarios, or whether it would just result in people's apps being terminated by the browser.

Your understanding is still correct.  I'm keeping this PR green while we discuss (for testing) and also making sure the native build work can support similar changes.

**Comment by jeffhandley** [comment]
I pushed a merge with master that resolved the conflict introduced from #52303.

**Comment by jeffhandley** [comment]
I'm going to convert this PR to a draft for now, to be revisited again soon.

**Comment by jeffhandley** [comment]
Assigning to @GrabYourPitchforks to track the need of documenting guidance for 6.0 and our plan for 7.0.
